### ISSUE TITLE
Proposal: Make rate limit optional

### ISSUE THREAD

**Issue by philippgille** [tsenart/vegeta#378]
#### Proposal

It seems like Vegeta requires a request rate. It must either be specified with `-rate` or a default rate of 1 per worker per second is used. I propose making this optional, leading to each worker making another request as soon as the previous one finished.
This is related to, but in my opinion different from https://github.com/tsenart/vegeta/issues/163 and https://github.com/tsenart/vegeta/issues/271.

#### Background

I'd like to know how many requests per second my web service supports at a reasonable response time.

Ramping up the rate (e.g. attacking for 5s with 50 requests second, then for 5 seconds with 100 requests per second etc.) makes it possible to "guess" the maximum request rate, because at some point the response times suddenly take much longer than before, but this is a little cumbersome to analyze to get precise numbers.

To expand on the previous paragraph regarding what I mean with "guessing" the maximum request rate: Let's say a rate of 50 leads to 30ms avg response time, a rate of 100 leads to 35ms avg response time, a rate of 200 leads to 50ms avg response time, then suddenly a rate of 400 leads to 5s response time. I would guess that the web service supports around 300 requests per second (overall, with 50 clients). To get a more precise number the increase of the rate can now be lowered and another round of load testing with rates between 200 and 400 requests is required.

I think this number can be determined easier and more precisely with the proposed approach.

#### Workarounds

Other load testing tools seem to be focused on exactly this approach, one of them being https://github.com/rakyll/hey.
Example: `hey -z 10s http://localhost:8080`
The first part of the result is:
```
Summary:
  Total:        10.0065 secs
  Slowest:      0.0237 secs
  Fastest:      0.0003 secs
  Average:      0.0065 secs
  Requests/sec: 7646.6975
...(snip)
```
Where `Requests/sec` is the number I'm interested in.

But I think in my specific case (a web service with large response bodies) the results I get from `hey` are off and Vegeta also seems to be more actively maintained, which is why I would like to see this feature in Vegeta.

**Comment by philippgille** [comment]
Similar: "Apache HTTP server benchmarking tool", a.k.a. `ab` (https://httpd.apache.org/docs/2.4/programs/ab.html).

Example:

```
$ ab -c 50 -t 10 "http://localhost:8080"
This is ApacheBench, Version 2.3 <$Revision: 1843412 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 5000 requests
...(snip)
Finished 31172 requests

Server Software:
...(snip)

Concurrency Level:      50
Time taken for tests:   10.001 seconds
Complete requests:      31172
Failed requests:        0
Total transferred:      1413198063 bytes
HTML transferred:       1409982712 bytes
Requests per second:    3116.89 [#/sec] (mean)
Time per request:       16.042 [ms] (mean)
Time per request:       0.321 [ms] (mean, across all concurrent requests)
Transfer rate:          137993.95 [Kbytes/sec] received
...(snip)
```
Here it's `Requests per second`.

**Comment by tsenart** [comment]
I have previously talked about this difference between load testing and benchmarking. As I once described it, benchmarking tries to find the limits of your system at its peak capacity, while load testing informs you how your system behaves at different load points and configurations.

With these definitions, it seems like benchmarking would be just another feature of a load testing solution, but I never came around to implement this in Vegeta.

In practice, a single isolated invocation of Vegeta (one process) won't be able to achieve the necessary requests rates to saturate many systems out there. It wouldn't, for instance, be able to find the breaking point of a local nginx instance responding with a static file. For that kind of benchmarking, I've been pointing people to other tools that choose different trade-offs:

- https://github.com/giltene/wrk2

I'm not opposed to extend Vegeta with benchmarking abilities, but that won't be an easy endeavour beyond the ~20k requests per second threshold. Additionally, I'm currently time constrained enough that I'm hesitant to commit to such a mission.

**Comment by briankung** [comment]
Sorry to resurrect an old thread, but:

> I have previously talked about this difference between load testing and benchmarking.

Do you have any links to prior discussions about this? Some searches of the issues and general google searching weren't particularly helpful.

**Comment by tsenart** [comment]
> Do you have any links to prior discussions about this? Some searches of the issues and general google searching weren't particularly helpful.

* Benchmarking: Finding the maximum limits of your system under test.
* Load testing: Understanding the performance of your system at different load points.

**Comment by tsenart** [comment]
FYI: I have a PR open for review that addresses this in some form. You still need to find your max capacity by playing around with the number of max-workers, but each worker will send requests as fast as it can, waiting for each response before sending the next one.

After this, I'll be working on a `-find` flag to the attack command where you can specify a latency metric threshold (e.g. 99th percentile <= 100ms), and vegeta will find the maximum `-rate` that maintains that condition / SLA for the specified duration.

**Comment by tsenart** [comment]
Hmmm, perhaps an `-sla` flag would be more suitable?

**Comment by tsenart** [comment]
Or maybe a combination of: `-rate=0` + [`-ramp`](#418) + `-until="99p > 100ms"`.

**Comment by jtlapp** [comment]
I really tried to make `wrk2` work for benchmarking, but I needed Deno on the same image. Deno requires recent Linux binaries, while `wrk2` requires older Linux binaries. After an entire weekend trying to resolve this conflict, I gave up and cast around for another option, ultimately deciding to see what I could do with Vegeta.

I have no prior experience benchmarking servers, but I'm doing a binary search for the highest rate that an API server can handle without errors or timeouts. I'm having to wait a long time between runs to ensure ports become available and exit TIME_WAIT. Find a rate is time-consuming, and the resulting rates seem to vary by 10% for any scenario.

I've wrapped the Vegeta library in my own program. Vegeta is very nice to work with. Thanks so much!
